sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/DeltaMetricStorage.java

```diff
@@ -96,7 +104,19 @@ public BoundStorageHandle bind(Attributes attributes) {
    * <p>All synchronous handles will be collected + reset during this method. Additionally cleanup
    * related stale concurrent-map handles will occur. Any {@code null} measurements are ignored.
    */
-  private synchronized void collectSynchronousDeltaAccumulationAndReset() {
+  private synchronized void collectSynchronousDeltaAccumulationAndReset(long epochNanos) {
+    // First check for a clock reset, and make sure we can limp along.
```

Comment:
I'm curious how this can end up happening. I thought that the nano clock was guaranteed to be monotonically increasing, and since our clocks are pinned to creation millis + monotonic nanos, I would think this couldn't happen. 

Do you have a scenario that was uncovered in OC where this was violated?

Reply:
Here's one of the original issues: https://github.com/census-instrumentation/opencensus-java/issues/2068
 
 TL;DR; I was unable to reproduce, and while it "should not happen", I'd rather be safe.
 
 Specifically for the nano-clock used in OTel I'm not sure if this will apply, but I also think we don't need to take chances.   IIRC This is one of a very few set of locations where timestamps matter in our metric collection vs. OpenCensus where there was the disruptor between measruement collection and aggregation.

Reply:
I suppose someone could implement a very badly behaved clock, since they can provide their own. Can you add a comment in here about the unlikely/impossibility of a clock reset when using the standard Clock implementation?

Reply:
Oops seemed to have missed hitting publish so affects you comment @jkwatson. I don't think we should use Clock or anything tied to wall time here

Reply:
@anuraaga one of my goals is to avoid hitting any kind of clock repeatedly once collection starts, i.e. we peg it once and then use a value throughout.

It it makes sense to not used anchored clock for this, I can swap to using some other time that is also passed down, however I'd like to avoid doing a clock-lookup per-metric per-instrumentation-library, as would be the case here.

Additionally, when it comes to Exemplar time measurement, it sounds like that should ALSO not use Anchored clock, so there's a bit of a time cleanup to occur here.

Going to open a thread on slack to ask some questions.

Reply:
Added #3691 to handle exemplars going forward *and* updated exemplars to call `.now` in this PR.

Still trying to understand the asks here on the slack channel.   The timestamp I'm using here is not associated with an AnchroedClock and I'm not sure what to anchor against to resolve this.   The timestamp is a wall-clock timestamp relating to the metric collection interval (start/current timestamps for aggregation temporality + gauges).  I built this logic trying to avoid adding a new clock and just re-using the existing lookup per-collection.

Reply:
Left some notes in Slack but adding here too.

I singled out Exemplar having forgotten that normal metrics report a wall time too. Anytime a wall time is needed, indeed we use wall time - `Clock.getDefault().now()` seems like a reasonable source for it. For Exemplars, using the span's anchored clock would prevent a time jump from causing the Exemplar to report a time outside a span's window - this may not be a big enough issue on the backend to care about it too much.

But anywhere that duration is needed, we should always use `System.nanoTime()` to avoid invalid durations due to time jump - this is a general best practice  that I'd like to not create exceptions for. This collection interval is a duration, so it should use monotonic time, i.e. `System.nanoTime()`, not wall time. I don't think using `epochNanos`, which is a wall time, is appropriate here based on the principal of "never use wall time for a duration", a valuable principal to hold.

Reply:
Thanks for copying slack convo here.  My next steps:

- I'm going to create a solution in SdkMetricProvider that does duration tracking using system nano time.
- A flag will be passed into metric storage denoting whether or not synchronous collection is needed, or we can just use the current reporting values.
- I'm going to fix the Clock.nanos() vs. Clock.now() call for exemplars in this PR (if it's ok).

Effectively after this PR, any area where we need wall-time in metrics SHOULD be using Clock.now(), and any location that needs duration will use Clock.nanos().

Reply:
Completely aside from this discussion, I think all of this is why it is very important that we figure out a first class Timer implementation for otel metrics, so our users don't have to worry about these details. :)

Reply:
> Effectively after this PR, any area where we need wall-time in metrics SHOULD be using Clock.now(), and any location that needs duration will use Clock.nanos().

Just want to confirm whether you're planning on doing it in this PR #3689 or a new one to follow up?